modules/compiler/utils/source/barrier_regions.cpp

```diff
@@ -1120,8 +1120,22 @@ Function *compiler::utils::Barrier::GenerateNewKernel(BarrierRegion &region) {
       // Check to see if it is still an alloca after vmap. If not we may
       // have processed it before and no work needs doing as we are using
       // the live variable struct directly.
-      if ((alloca_inst = dyn_cast<AllocaInst>(vmap[alloca_inst]))) {
-        allocas_to_remove.push_back(alloca_inst);
+      if (auto *new_alloca_inst = dyn_cast<AllocaInst>(vmap[alloca_inst])) {
+        allocas_and_intrinsics_to_remove.push_back(new_alloca_inst);
+        // Also remove any assume-like intrinsics that are users of this
+        // alloca. These assumptions may not hold. For example, lifetime
+        // intrinsics are definitely dangerous, as by directly replacing their
+        // alloca operands with the address of the live variable struct, we are
+        // telling LLVM that *all* accesses of the live variable struct also
+        // start/end at that point, which is not true.
+        // Similarly, llvm.assume and llvm.experimental.noalias.scope.decl may
+        // hold for the alloca but not the live variables struct.
+        for (auto *const user : alloca_inst->users()) {
+          if (auto *const lifetime = dyn_cast<IntrinsicInst>(user);
```

Comment:
presumably this could catch no-lifetime intrinsics? If so, the name is misleading

Reply:
whoops yeah, I originally started out handling only lifetime intrinsics but saw that we were buggy in other ways. I've updated to simply `intrinsic`